File: Python WorldSimulatorGame/Roslina.py
from Organizm import Organizm
from Punkt import Punkt
import random


class Roslina(Organizm):

    def __init__(self, sila, inicjatywa, znak, wiek, zycie, swiat, polozenie):
        super(Roslina, self).__init__(sila, inicjatywa, znak, wiek, zycie, swiat, polozenie)

    def akcja(self):
        self.wiek += 1
        self.ogolneInformacje()
        x = random.randint(0, 9)
        if x == 0:
            self.rozprzestrzenianie()

    def kolizja(self, other: Organizm, atakujacy, ruch):
        self.zabojczeInformacjeDlaRoslin(other, self)
        # print(" " + other.getNazwaOrganizmu() + " zjadl -> " + self.getNazwaOrganizmu())
        self.setZycie(False)

    def rozprzestrzenianie(self):
        miejsceUrodzenia = self.znajdzWolnyPunktObok(self.polozenie)

        if self.polozenie.sprwadzCzyTenSam(miejsceUrodzenia) is False:
            self.swiat.dodajOrganizm(self.getZnak(), miejsceUrodzenia)
            #print(" Rozprzestrzenia sie! Powstaje nowa " + self.getNazwaOrganizmu() + " na polu: " + miejsceUrodzenia.printPunkt(), end='')
            self.swiat.dodajInformacje(" Rozprzestrzenia sie! Powstaje nowa " + self.getNazwaOrganizmu() + " na polu: " + miejsceUrodzenia.printPunkt())

    def znajdzWolnyPunktObok(self, stary: Punkt):

        nowy = Punkt()
        if stary.getY() != 0 and self.swiat.sprawdzCzyPolePuste(stary.getX(), stary.getY() - 1):
            nowy.setX(stary.getX())
            nowy.setY(stary.getY() - 1)
        elif stary.getY() != self.swiat.getHeight() - 1 and self.swiat.sprawdzCzyPolePuste(stary.getX(),
                                                                                           stary.getY() + 1):
            nowy.setX(stary.getX())
            nowy.setY(stary.getY() + 1)
        elif stary.getX() != 0 and self.swiat.sprawdzCzyPolePuste(stary.getX() - 1, stary.getY()):
            nowy.setX(stary.getX() - 1)
            nowy.setY(stary.getY())
        elif stary.getX() != self.swiat.getWidth() - 1 and self.swiat.sprawdzCzyPolePuste(stary.getX() + 1,
                                                                                          stary.getY()):
            nowy.setX(stary.getX() + 1)
            nowy.setY(stary.getY())
        else:
            nowy = stary
        return nowy

File: C++ WorldSimulatorGame/POproj1/Roslina.cpp
#include "Roslina.h"
#include <windows.h>

Roslina::Roslina(int sila, int inicjatywa, char znak, int wiek, bool zycie, Organizmy organizm, Punkt polozenie, Swiat* swiat)
	:Organizm(sila, inicjatywa, znak, wiek, zycie, organizm, polozenie, swiat) {

}

void Roslina::akcja() {
	this->wiek++;
	ogolneInformacje();
	//cout << this->getZnak() << " ";
	//cout << " Ma tyle sily: " << this->getSila() << endl;
	//cout << " Z pola: " << this->polozenie.getX() << " " << this->polozenie.getY();
	int x = rand() % 10;
	if (x < 1) {
		Rozprzestrzenianie();
	}
	else {
		//cout << " Nie udalo sie rozprzestrzenic" << endl;
		swiat->info.dodajInfocjmacje(" Nie udalo sie rozprzestrzenic ");
	}
	swiat->info.dodajInfocjmacje("\n");
}
void Roslina::kolizja(Organizm* other, bool atakujacy, bool* ruch) {

	zabojczeInformacjeDlaRoslin(other);
	//std::cout << other->getZnak() << " z sila: " << other->getSila() << " zjadl -> " << this->getZnak() << " z sila: " << this->getSila() << std::endl;
	swiat->wyczyscOrganizmZMapy(this->polozenie);
}

void Roslina::Rozprzestrzenianie() {
	Punkt p2 = znajdzWolnyPunktObok(this->polozenie);
	if (p2 != this->getPolozenie()) {
		swiat->info.dodajInfocjmacje("Rozprzestrzenia sie! Powstaje nowa roslina na polu : " + to_string(p2.getX()) + " " + to_string(p2.getY()));
		//cout << " Powstaje nowa roslina na polu: " << p2.getX() << " " << p2.getY() << endl;
		swiat->stworzRosline(p2, this->getZnak());
	}
	else {
		swiat->info.dodajInfocjmacje(" Nie udalo sie rozprzestrzenic ");
	}
}
Punkt Roslina::znajdzWolnyPunktObok(Punkt p) {
	Punkt p2 = p;
	if (p.getY() != 0 && swiat->czyPustePole(p.getX(), p.getY() - 1)) {
		p2.setY(p.getY() - 1);
		p2.setX(p.getX());
	}
	else if (p.getY() != swiat->getHeight() - 1 && swiat->czyPustePole(p.getX(), p.getY() + 1)) {
		p2.setY(p.getY() + 1);
		p2.setX(p.getX());

	}
	else if (p.getX() != 0 && swiat->czyPustePole(p.getX() - 1, p.getY())) {
		p2.setY(p.getY());
		p2.setX(p.getX() - 1);

	}
	else if (p.getX() != swiat->getWidth() - 1 && swiat->czyPustePole(p.getX() + 1, p.getY())) {
		p2.setY(p.getY());
		p2.setX(p.getX() + 1);

	}
	return p2;
}

void Organizm::rysowanie() {
	HANDLE hConsole = GetStdHandle(STD_OUTPUT_HANDLE);
	SetConsoleTextAttribute(hConsole, 10);
	cout << this->znak;
	SetConsoleTextAttribute(hConsole, 7);
}

Roslina::~Roslina() {
}